## Utilizacion de Xi cuadrado y t-Test 



* PARA ESTIMAR LA RELACION ENTRE VARIABLES CUALITATIVAS (chi-test) Y DE VARIABLES CUANTITATIVAS CON CUALITATIVAS (t-test)

Dado que la correlacion solo nos calcula para variables cuantitativas, tanto discretas como continuas, pero no para cualitativas. Solo sirven para numericas.

ESTE ANALISIS HA DE HACERSE PREVIO A SELECCIONAR VARIABLES PREDICTORAS CUALITATIVAS PARA CREAR UN MODELO O CUANDO LA VARIABLE OBJETIVO DEL MODELO ES CUALITATIVA.

* Chi cuadrado

Es un procedimiento estadístico utilizado para examinar las diferencias entre variables categóricas en la misma población.

Prueba de chi-cuadrado(de Pearson o prueba exacta de Fisher) para la independencia: le permite probar si existe o no una asociación estadísticamente significativa entre dos variables categóricas. 


* T-test

Del mismo modo que Chi cuadrado, permite probar si existe relacion entre una variable categorica y otra cuantitativa.

Se usa para averiguar si hay una diferencia real entre las medias (promedios) de dos grupos diferentes. 

###### Estos test plantean dos hipotesis, H0 las variables son independientes y H1 las variables estan relacionadas.

Cuando rechaza la hipótesis nula de una prueba de chi-cuadrado o de t-test para la independencia, significa que hay una asociación significativa entre las dos variables.

* La forma de interpretar la salida de los test es la siguiente:


Las pruebas de independencia de chi-cuadrado y t-test utilizan las siguientes hipótesis: nulas y alternativa

    - H 0 : (hipótesis nula) Las dos variables son independientes. La aceptamos cuando p-valor > umbral (0,05)
    - H 1 : (hipótesis alternativa) Las dos variables son dependientes. La aceptamos cuando p-valor < umbral (0,05)




In [2]:
import pandas as pd
import numpy as np
import random
from numpy.random import seed
from numpy.random import randn
from scipy.stats import t
from scipy.stats import chi2
from scipy.stats import ttest_ind
import scipy.stats as stats

In [3]:
import scipy

### T-test para variables cualitativa vs cuantitativa

Peso ~ genero

In [4]:
seed=1987
df=pd.DataFrame({'mujer':np.random.randint(10,100, size=10), 'hombre':np.random.randint(10,150,size=10)})
df

# Las columnas son los grupos por variable cualitativa

# Tenemos 20 personas, 10 en cada grupo. Esta compensada

mujer  hombre
0     89      31
1     42      42
2     54     137
3     11     135
4     62      47
5     67     148
6     97     105
7     39      73
8     58      52
9     77      73

In [5]:
ttest_ind(df['mujer'],df['hombre'])

Ttest_indResult(statistic=-1.5481068215082212, pvalue=0.1389987346540685)

In [6]:
riesgo=0.05
p_value=ttest_ind(df['mujer'],df['hombre']) # variable trim para indicar el nivel de riesgo por defecto 0,05
print (p_value)
print('\nP-valor',p_value[1],'\n')
if p_value[1]>riesgo:
    print('Aceptamos H0: Las variables son independientes')
else:
    print('Rechazamos H0, aceptamos H1: Las variables tienen dependencia')

Ttest_indResult(statistic=-1.5481068215082212, pvalue=0.1389987346540685)

P-valor 0.1389987346540685 

Aceptamos H0: Las variables son independientes


##### Otro ejemplo de t-test

In [7]:
# Metodos de estudio y resultados obtenidos

df = pd.DataFrame({'method': ['A', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'A',
                              'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B'],
                   'score': [71, 72, 72, 75, 78, 81, 82, 83, 89, 91, 80, 81, 81,
                             84, 88, 88, 89, 90, 90, 91]})

In [8]:
df.head(3)

method  score
0      A     71
1      A     72
2      A     72

In [11]:
groupA = df[df['method']=='A']
groupB = df[df['method']=='B']

In [14]:
groupB

method  score
10      B     80
11      B     81
12      B     81
13      B     84
14      B     88
15      B     88
16      B     89
17      B     90
18      B     90
19      B     91

In [51]:
ttest_ind(groupA['score'], groupB['score'])

Ttest_indResult(statistic=-2.6034304605397938, pvalue=0.017969284594810425)

O bien

In [52]:
ttest_ind(df[df['method']=='A']['score'], df[df['method']=='B']['score']) 

# se le han de pasar los valores de cada grupo poblacional

Ttest_indResult(statistic=-2.6034304605397938, pvalue=0.017969284594810425)

Resultado:

Estadística de la prueba T: –2,6034

Valor P: 0.0179

Dado que el valor p es menor que .05, rechazamos la hipótesis nula y concluimos que hay evidencia suficiente para decir que los dos métodos conducen a diferentes puntajes medios en el examen, o sea que las variables estan vinculadas.

### Chi-test para variables cualitativas 

* X e Y son cualitativas. Necesitamos crear una tabla de contingencia (frecuencia)

In [17]:
np.random.seed(1987)
tipo_botella=np.random.choice(['papel','vidrio','carton','plastico','otro'],p=[0.05,0.15,0.25,0.05,0.5], size=1000)
mes=np.random.choice(a=['Enero','Febrero','Marzo'], p=[0.4,0.2,0.4], size = 1000)
botellas=pd.DataFrame({'Tipo':tipo_botella, 'Produccion': mes})
botellas

Tipo Produccion
0      carton      Marzo
1      vidrio      Enero
2        otro      Marzo
3        otro      Marzo
4      carton      Marzo
..        ...        ...
995  plastico      Enero
996      otro      Enero
997      otro    Febrero
998    carton    Febrero
999      otro    Febrero

[1000 rows x 2 columns]

* Necesitamos la tabla de contigencia con el recuento para cada cruce de las dos variables

In [18]:
tabla_contingencia=pd.crosstab(botellas.Tipo, botellas.Produccion, margins=True)

In [19]:
tabla_contingencia

Produccion  Enero  Febrero  Marzo   All
Tipo                                   
carton        113       40    105   258
otro          226       86    197   509
papel          11        9     17    37
plastico       16       11     23    50
vidrio         63       22     61   146
All           429      168    403  1000

In [21]:
observados=tabla_contingencia.iloc[0:5,0:3] # quito los totales
observados

Produccion  Enero  Febrero  Marzo
Tipo                             
carton        113       40    105
otro          226       86    197
papel          11        9     17
plastico       16       11     23
vidrio         63       22     61

In [55]:
pd.crosstab(botellas.Tipo, botellas.Produccion, margins=False) # si pongo margins=False ya no necesito quitar los totales

Produccion  Enero  Febrero  Marzo
Tipo                             
carton        113       40    105
otro          226       86    197
papel          11        9     17
plastico       16       11     23
vidrio         63       22     61

In [28]:
scipy.stats.chi2_contingency(observed=observados)

(6.878957335826994,
 0.5497469290521917,
 8,
 array([[110.682,  43.344, 103.974],
        [218.361,  85.512, 205.127],
        [ 15.873,   6.216,  14.911],
        [ 21.45 ,   8.4  ,  20.15 ],
        [ 62.634,  24.528,  58.838]]))

In [29]:
p_value=scipy.stats.chi2_contingency(observed=observados)
p_value[1] # es un valor alto (>0,05 nivel de riesgo)

0.5497469290521917

In [34]:
riesgo=0.05
p_value = scipy.stats.chi2_contingency(observed=observados) # variable trim para indicar el nivel de riesgo por defecto 0,05
print (p_value)
print('\nP-valor',p_value[1],'\n')
if p_value[1]>riesgo:
    print('Aceptamos H0: Las variables son independientes')
else:
    print('Rechazamos H0, aceptamos H1: Las variables tienen dependencia')

(6.878957335826994, 0.5497469290521917, 8, array([[110.682,  43.344, 103.974],
       [218.361,  85.512, 205.127],
       [ 15.873,   6.216,  14.911],
       [ 21.45 ,   8.4  ,  20.15 ],
       [ 62.634,  24.528,  58.838]]))

P-valor 0.5497469290521917 

Aceptamos H0: Las variables son independientes


##### Otro Ejemplo de Xi cuadrado

In [54]:
# tabla de frecuencias genero (filas) X Partido politico (columnas) 

datos = [[120, 90, 40],
        [110, 95, 45]]

#Cuando tenemos los datos en un df debemos de obtener a partir de los datos la tabla de contingencia o frecuencias. 
#Aqui ya la tenemos conformada


In [42]:
# prueba chi-cuadrado de independencia
scipy.stats.chi2_contingency(datos)


(0.8640353908896108,
 0.6491978887380976,
 2,
 array([[115. ,  92.5,  42.5],
        [115. ,  92.5,  42.5]]))

In [43]:
#Estadística de prueba de chi-cuadrado: 0.864

# p-valor: 0,649

#Grados de libertad: 2 (calculado como # filas-1 * # columnas-1)

#Matriz: la última matriz muestra los valores esperados para cada celda en la tabla de contingencia.


Dado que el p-valor (0.649) de la prueba NO es menor que 0.05, no rechazamos la hipótesis nula.

Las preferencias de género y de partido político son independientes.